src/headless/plugins/disco/entity.js

```diff
@@ -40,6 +40,13 @@ const DiscoEntity = Model.extend({
         id = `converse.disco-items-${this.get('jid')}`;
         this.items.browserStorage = _converse.createStore(id, 'session');
         await new Promise(f => this.items.fetch({'success': f, 'error': f}));
+        _converse.disco_entities
+            .filter(e => {
+                // XXX: Ignore nodes for now.
+                // See: https://xmpp.org/extensions/xep-0030.html#items-nodes
+                return typeof e.get('name') !== 'undefined';
+            })
+            .forEach(e => this.items.create(e));
```

Comment:
@SilverYoCha I applaud your tenacity and dedication to finding this bug :) :clap: 

That said, I don't think this change itself is appropriate, it's a plaster over an underlying issue.

Your change here gave me a hunch as to what the issue is, and so I did some digging.

The fact that the icon consistently showed for me on Prosody but not for you on Ejabberd is to me an indication that this was due to some kind of race condition.

The change you made above avoided the race condition by immediately creating all global entities inside this entity. This is BTW wrong, since you're now letting this entity state that is has as `#items` all the entities of the global `_converse.disco_entities` collection and by extension that it supports the features of all those entities.
You would have to modify the code to first check whether the entity you're adding belongs to this current `DiscoEntity` as a sub-item and AFAIK there's no way currently to do that.

In any case, I managed to construct a test that reproduced the toolbar button not showing, even though the disco response indicated that it's supported.

Here's the fix: https://github.com/conversejs/converse.js/commit/a251608fc50390c7f46ae6ffbfbe9cd9774c8687

I'm hopeful that fix will now finally put this issue to bed.

Can you please check?

Reply:
Thank you @jcbrand for your comments and for taking my suggestion into consideration.

My correction was indeed more of a band-aid than anything else. I lack a lot of knowledge in the project to be effective.

I'm glad I could put you on a serious track. Unfortunately, your commit did not solve the problem in my environment. Right after login, the download button is displayed as expected (and as it was already), but after a CTRL+F5, it is not displayed anymore.

I noticed something strange from the beginning: is it normal that I never find an entry in the session storage with the prefix id `converse.disco-items-`?

Thank you again.

Reply:
Hi @SilverYoCha 

You have already been very effective and helped move this issue forwards.

It's taken me a long time to get an idea of what's going on because it's a complex/tricky bug, I cannot reproduce the issue with my XMPP server,  and I haven't had time to sit down and spend a long time to focus on figuring it out.

Based on what you've said before and your original MR, my current theory is that `this.items` is always empty because the entities are shared between two collections and for some reason are then persisted only to the one collection (`_converse.disco_items`) and not the other one (e.g. the MUC's entity).

I tried to verify this now, but with Prosody, the `Strophe.NS.HTTPUPLOAD` feature is on the MUC entity itself, not on one of the sub-items. I *think*, but haven't verified, that with Ejabberd, the feature is on a sub-item, and that this is the fundamental difference that is causing the bug to happen with Ejabberd and not Prosody. So with Prosody, it doesn't matter that `this.items` is always empty after a page reload because with Prosody the feature is on the MUC entity itself, not on a sub-entity.

I'm partly writing this down to also help myself remember. My day is full and I usually have 30 minutes to look at something like this before I have to do something else.

If I could get an Ejabberd account somewhere, I could verify this theory. I've asked in the Converse MUC if someone can help. I don't have time to set everything up myself.

> I noticed something strange from the beginning: is it normal that I never find an entry in the session storage with the prefix id converse.disco-items-?

I just checked, and I do see them in sessionStorage. They look like this: `converse-session/converse.disco-entities-XXXX`

You should be having entries like that as well.

![Screenshot from 2022-11-10 08-33-02](https://user-images.githubusercontent.com/683911/201027922-11e46e3f-901e-4b8a-b06e-f335ef84b2db.png)

Reply:
If it helps, I have created an account in my environment for you @jcbrand.
I will email you the conversejs initialization settings that I use along with your credentials.